Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -1047,9 +1049,134 @@ public class TypeArgumentInstantiation {
     protected abstract ILvalue EmitMemberSelect(Action<ConcreteSyntaxTree> obj, Type objType, MemberDecl member, List<TypeArgumentInstantiation> typeArgs, Dictionary<TypeParameter, Type> typeMap,
       Type expectedType, string/*?*/ additionalCustomParameter = null, bool internalAccess = false);
 
-    protected void EmitArraySelect(string index, Type elmtType, ConcreteSyntaxTree wr) {
-      EmitArraySelect(new List<string>() { index }, elmtType, wr);
+    protected enum MemberCompileStatus { Ordinary, Identity, AlwaysTrue }
+
+    protected MemberCompileStatus GetMemberStatus(MemberDecl member) {
```

Comment:
I would move the new members of `SinglePassCompiler`, GetMemberStatus, IsInvisibleWrapper, CompiledTypes and SimplifyType to a separate class with a name like `WrapperTypeEraser`.

Reply:
I moved them. The way I did it, the qualified names now contain some redundancy.